gateway/api.go

```diff
@@ -1111,6 +1169,82 @@ func (gw *Gateway) apiHandler(w http.ResponseWriter, r *http.Request) {
 	doJSONWrite(w, code, obj)
 }
 
+func (gw *Gateway) apiOASGetHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		apiID       = mux.Vars(r)["apiID"]
+		scopePublic = r.URL.Query().Get("mode") == "public"
+		obj         interface{}
+		code        int
+	)
+	if apiID != "" {
+		log.Debugf("Requesting API definition for %q", apiID)
+		obj, code = gw.handleGetAPIOAS(apiID, scopePublic)
+	} else {
+		log.Debug("Requesting API list")
+		obj, code = gw.handleGetAPIListOAS(scopePublic)
+	}
+
+	doJSONWrite(w, code, obj)
+}
+
+func (gw *Gateway) apiOASPostHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		obj  interface{}
+		code int
+	)
+
+	log.Debug("Creating new definition file")
+	obj, code = gw.handleAddOrUpdateApi("", r, afero.NewOsFs(), true)
```

Comment:
The request is passed because of the body. If that is the case, and that's the only use of *Request, then pass r.Body directly as io.Reader?

Can you split the function into handleAddAPI and handleUpdateApi? The code will be clearer to read, no need to pass empty arguments. Also see comment about typed arguments *struct instead of *http.Request